pattern/equivalence/AlphaEquivalence.java

```diff
@@ -17,262 +17,115 @@
-            return AlphaEquivalence.invalid();
-        }
-
-        @Override
-        public Valid addMapping(Variable from, Variable to) {
-            map.put(from, to);
-            reverseMap.put(to, from);
-            assert from.reference().isName() == to.reference().isName();
-            return this;
-        }
-
-        @Override
-        public AlphaEquivalence validIf(boolean valid) {
-            if (!valid) return AlphaEquivalence.invalid();
-            return this;
-        }
-
-        @Override
-        public boolean isValid() {
-            return true;
-        }
-
-        @Override
-        public Valid asValid() {
-            return this;
-        }
-
-        public Map<Retrievable, Retrievable> idMapping() {
-            return variableMapping().entrySet().stream()
-                    .map(e -> new AbstractMap.SimpleEntry<>(
-                            e.getKey().id(),
-                            e.getValue().id()))
-                    .filter(e -> {
-                        assert e.getKey().isRetrievable() == e.getValue().isRetrievable();
-                        return e.getKey().isRetrievable();
-                    })
-                    .collect(Collectors.toMap(
-                            e -> e.getKey().asRetrievable(),
-                            e -> e.getValue().asRetrievable()
-                    ));
-        }
-
-        Map<Variable, Variable> variableMapping() {
-            return new HashMap<>(map);
-        }
-
-        private Map<Variable, Variable> reverseVariableMapping() {
-            return new HashMap<>(reverseMap);
-        }
+        return Optional.of(existing);
     }
 
-    public static class Invalid extends AlphaEquivalence {
-
-        @Override
-        public Invalid addOrInvalidate(AlphaEquivalence mapping) {
-            return this;
-        }
-
-        @Override
-        public AlphaEquivalence addOrInvalidate(Supplier<AlphaEquivalence> mappingSupplier) {
-            return this;
-        }
-
-        @Override
-        public <T extends AlphaEquivalent<T>> Invalid validIfAlphaEqual(T member1, T member2) {
-            return this;
-        }
-
-        @Override
-        public Invalid addMapping(Variable from, Variable to) {
-            return this;
-        }
-
-        @Override
-        public Invalid validIf(boolean invalidate) {
-            return this;
-        }
-
-        @Override
-        public boolean isValid() {
-            return false;
-        }
-
-        @Override
-        public Valid asValid() {
-            throw new ClassCastException("Cannot cast " + this.getClass().getSimpleName() + " to " + Valid.class.getSimpleName());
-        }
-
-        @Override
-        public <T extends AlphaEquivalent<T>> AlphaEquivalence validIfAlphaEqual(Set<T> set1, Set<T> set2) {
-            return this;
-        }
+    public AlphaEquivalence extend(Variable from, Variable to) {
+        assert from.reference().isName() == to.reference().isName();
```

Comment:
ah we've not been good about this in reasoner but i think we should be using `from.id()` instead of `from.reference()` -- actually I think we should remove `reference()` from the API at all